src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -529,8 +585,60 @@ void GcsTaskManager::OnJobFinished(const JobID &job_id, int64_t job_finish_time_
         // workers associated with the job will be killed.
         task_event_storage_->MarkTasksFailedOnJobEnds(job_id,
                                                       job_finish_time_ms * 1000 * 1000);
+
+        // Clear and summarize the job summary info (since it's now finalized).
+        task_event_storage_->UpdateJobSummaryOnJobDone(job_id);
       });
 }
 
+void GcsTaskManager::GcsTaskManagerStorage::JobTaskSummary::GcOldDroppedTaskAttempts(
+    const JobID &job_id) {
+  const auto max_dropped_task_attempts_tracked_in_gcs = static_cast<size_t>(
+      RayConfig::instance()
+          .task_events_max_dropped_task_attempts_tracked_per_job_in_gcs());
+  if (dropped_task_attempts_.size() <= max_dropped_task_attempts_tracked_in_gcs) {
+    return;
+  }
+  RAY_LOG(INFO) << "Evict extra dropped task attempts(" << dropped_task_attempts_.size()
+                << " > " << max_dropped_task_attempts_tracked_in_gcs
+                << ") tracked in GCS for job=" << job_id.Hex() << ". Setting the "
+                << "RAY_task_events_max_dropped_task_attempts_tracked_per_job_in_gcs"
+                << " to a higher value to store more.";
+
+  // Pass 1: GC any dropped task attempt if they have not been updated.
```

Comment:
hmm is this part necessary? Feel pretty expensive to have 1M iteration every 5 seconds. Maybe let's just remove it as this is a rare edge cases?

Reply:
So this is to NOT gc task attempts being dropped recently, which might have other task events coming in. 

If we don't do this, we might run into scenarios where we will run into partial loss more frequently. 

The average time for this is 35ms from the release test - I am open to remove this tho.

Reply:
synced offline - we will just do hard limit + 10%, and no more time based.